crates/api_common/src/websocket/chat_server.rs

```diff
@@ -0,0 +1,395 @@
+    }
+
+    // If the room doesn't exist yet
+    if inner.community_rooms.get_mut(&community_id).is_none() {
+      inner.community_rooms.insert(community_id, HashSet::new());
+    }
+
+    inner
+      .community_rooms
+      .get_mut(&community_id)
+      .context(location_info!())?
+      .insert(id);
+    Ok(())
+  }
+
+  pub fn join_mod_room(
+    &self,
+    community_id: CommunityId,
+    id: ConnectionId,
+  ) -> Result<(), LemmyError> {
+    let mut inner = self.inner();
+    // remove session from all rooms
+    for sessions in inner.mod_rooms.values_mut() {
+      sessions.remove(&id);
+    }
+
+    // If the room doesn't exist yet
+    if inner.mod_rooms.get_mut(&community_id).is_none() {
+      inner.mod_rooms.insert(community_id, HashSet::new());
+    }
+
+    inner
+      .mod_rooms
+      .get_mut(&community_id)
+      .context(location_info!())?
+      .insert(id);
+    Ok(())
+  }
+
+  pub fn join_post_room(&self, post_id: PostId, id: ConnectionId) -> Result<(), LemmyError> {
+    let mut inner = self.inner();
+    // remove session from all rooms
+    for sessions in inner.post_rooms.values_mut() {
+      sessions.remove(&id);
+    }
+
+    // Also leave all communities
+    // This avoids double messages
+    // TODO found a bug, whereby community messages like
+    // delete and remove aren't sent, because
+    // you left the community room
+    for sessions in inner.community_rooms.values_mut() {
+      sessions.remove(&id);
+    }
+
+    // If the room doesn't exist yet
+    if inner.post_rooms.get_mut(&post_id).is_none() {
+      inner.post_rooms.insert(post_id, HashSet::new());
+    }
+
+    inner
+      .post_rooms
+      .get_mut(&post_id)
+      .context(location_info!())?
+      .insert(id);
+
+    Ok(())
+  }
+
+  pub fn join_user_room(&self, user_id: LocalUserId, id: ConnectionId) -> Result<(), LemmyError> {
+    let mut inner = self.inner();
+    // remove session from all rooms
+    for sessions in inner.user_rooms.values_mut() {
+      sessions.remove(&id);
+    }
+
+    // If the room doesn't exist yet
+    if inner.user_rooms.get_mut(&user_id).is_none() {
+      inner.user_rooms.insert(user_id, HashSet::new());
+    }
+
+    inner
+      .user_rooms
+      .get_mut(&user_id)
+      .context(location_info!())?
+      .insert(id);
+    Ok(())
+  }
+
+  async fn send_post_room_message<OP, Response>(
+    &self,
+    op: &OP,
+    response: &Response,
+    post_id: PostId,
+    websocket_id: Option<ConnectionId>,
+  ) -> Result<(), LemmyError>
+  where
+    OP: OperationType + ToString,
+    Response: Serialize,
+  {
+    let mut inner = self.inner();
+    let res_str = &serialize_websocket_message(op, response)?;
+    if let Some(sessions) = inner.post_rooms.get(&post_id).cloned() {
+      for id in sessions {
+        if let Some(my_id) = websocket_id {
+          if id == my_id {
+            continue;
+          }
+        }
+        inner.sessions.get_mut(&id).unwrap().text(res_str).await?;
+      }
+    }
```

Comment:
Im confused by this logic (which also fails clippy currently). It seems to send the message to everyone who is a member of the post room, with exception of the connection passed as `websocket_id`. Is that correct?

Reply:
I see, it seems to be because the user who made the api call already gets a response that way. Here the user needs to be filtered out to avoid duplicate notifications.